content/blog/2025-05-13-imperative-quantifier-elimination.md

```diff
@@ -0,0 +1,178 @@
+
+1. The `range` bounds are explicitly enforced: `any(... for x in range(y, z))` becomes `any(... and x >= y and x < z for x in range(y,z))`. This is redundant at this stage, but will be critical when we remove the `range` call in the next step.
+2. There are no `!=`, `x != y` is replaced with `x < y or y < x`
+3. There are no `<=` or `>=`, `x <= y` is replaced with `x < y + 1` (remember that `x` and `y` are both integers)
+4. There are no `==`, `any(x == <expr> and x < y for x in ...)` becomes `any(<expr> < y for x in ...)`
+5. The quantifier variable appears only on the left side of inequalities and no other variable appears on the left side. Furthermore, all inequalities are scaled so that the same multiple of the quantifier variable appears on the left side (i.e. if `x` is the quantifier variable, `x < y and 2*x < z` becomes `2*x < 2*y and 2*x < z`).
+
+### The Final Simple Pass
+
+At this point we have a very regular AST. It looks like this:
+
+```Python
+any(m*x < <expr> and m*x > <expr> and ... for x in range(...))
+```
+
+The key insight is that of all the upper bound sub-predicates of the form `m*x < <expr>`, only one of them will matter. Only one of them will be the _least upper bound_. Similarly, only the _greatest lower bound_ will actually restrict what values of `x` return `True`. We don't know statically which upper bound is the least upper bound, because the upper bounds may have variables in them, but we can always check at runtime.
+
+In a pure predicate logic setting, these special bounds can be "calculated" by statically unrolling a big loop that checks if each (lower bound, upper bound) pair is correct. But in Python we can just call `max` and `min`!
+
+With the least upper bound and the greatest lower bound in hand, one only needs to check if there is a value between them that is divisible by `m` (remember that `m*x` needs to satisfy all of these inequalities, not `x`) and the quantifier elimination is complete.
+
+# The Evaluation
+
+### Correctness
+
+Compiler passes (even those for class projects) should not change the behavior of the code they modify. This pass does not meet that standard for two reasons:
+
+1. As I mentioned above, this is a static transformation of untyped Python code. There is no guarantee that all of the variables in the program are integers or even that `+` does what I expect. So the best we can hope for is that it is correct _for integer inputs_.
+2. If I found all of the bugs in my implementation with testing I think that might be the first recorded instance of that happening in all of human history.
+
+That being said, I did some testing! Because this pass operates on such a simple sub-language of Python, I built a random program generator and used those programs as my test inputs. I generated around 60 programs and when one of them failed (my pass changed the program behavior) I would whittle it down to a minimal failing example and keep that as its own test.
```

Comment:
Really cool (and practical) approach to constructing tests. How many tests did you end up with after going through this process a few times?

Reply:
I added a note about this, but it was roughly 60 randomly generated programs, of which 5-ish failed and were minimized. Although at least one failing program triggered multiple bugs and had to be minimized more than once to different minimal failing examples.